localstack/extensions/patterns/webapp.py

```diff
@@ -0,0 +1,321 @@
+import importlib
+import logging
+import mimetypes
+import typing as t
+from functools import cached_property
+
+from rolo.gateway import HandlerChain
+from rolo.router import RuleAdapter, WithHost
+from werkzeug.routing import Submount
+
+from localstack import config
+from localstack.aws.api import RequestContext
+from localstack.extensions.api import Extension, http
+
+if t.TYPE_CHECKING:
+    # make sure jinja2 stays optional
+    import jinja2
+
+LOG = logging.getLogger(__name__)
+
+_default = object()
+
+
+class WebAppExtension(Extension):
+    """
+    EXPERIMENTAL! This class is experimental and the API may change without notice.
+
+    A webapp extension serves routes, templates, and static files via a submount and a subdomain through localstack.
+
+    It assumes you have the following directory layout::
+
+        my_extension
+        ├── extension.py
+        ├── __init__.py
+        ├── static              <-- make sure static resources get packaged!
+        │   ├── __init__.py
+        │   ├── favicon.ico
+        │   └── style.css
+        └── templates            <-- jinja2 templates
+            └── index.html
+
+    Given this layout, you can define your extensions in ``my_extension.extension`` like this. Routes defined in the
+    extension itself are automatically registered::
+
+        class MyExtension(WebAppExtension):
+            name = "my-extension"
+
+            @route("/")
+            def index(request: Request) -> Response:
+                # reference `static/style.css` to serve the static file from your package
+                return self.render_template_response("index.html")
+
+            @route("/hello")
+            def hello(request: Request):
+                return {"message": "Hello World!"}
+
+    This will create an extension that localstack serves via:
+
+    * Submount: https://localhost.localstack.cloud:4566/_extension/my-extension
```

Comment:
nit: I guess this already evolved and is used for existing extensions, which would make it hard to change, but I would suggest to use the plural for API paths that contain lists:
```suggestion
    * Submount: https://localhost.localstack.cloud:4566/_extensions/my-extension
```

And I guess it was already discussed in the past if the `_extension` endpoint should be a top-level entry / first class citizen in LocalStack? For example, CloudPods live in `/_localstack/pods`. But extensions can also be totally unrelated to LocalStack itself (like `mailhog`), where it might even be confusing...

Reply:
well it hasn't really been discussed directly, i'm making a "product owner" decision here to introduce it. i believe having a well-known namespace for extensions is important to have since it allows us to reason over the requests. moreover, it should be distinct from `/_localstack` for the reason you correctly identified. we should port our existing extensions to that convention.

i prefer singular for the following (completely arbitrary and personal taste reason which you are welcome to disagree with :D). we have `/_localstack/extensions` which indicates that you can _manage_ one or more extensions. whereas the singular `/_extension/<something>` indicates that you're talking to an individual and specific extension.

Reply:
I feel like it's a best practice when it comes to HTTP / RESTful API design to use the plural for path segments that in turn contains a collection of sub-resources (`/_extensions` is a collection resource whereas the specific extension route would expose a single resource). I don't see a reason to use the `/_extension(s)` path in the future to show metadata on the _collection_ of extensions registered in this "namespace", for example a `GET /_extensions` listing the different sub-routes for `WebAppExtensions`...
But again, just a nitpick 😛

Reply:
All fair points. This is not a REST API though :)